archivers/pax/files/pax.1
.\"    $NetBSD: pax.1,v 1.5 2005/12/01 03:00:01 minskim Exp $
.\"
.\" Copyright (c) 1992 Keith Muller.
.\" Copyright (c) 1992, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" This code is derived from software contributed to Berkeley by
.\" Keith Muller of the University of California, San Diego.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)pax.1	8.4 (Berkeley) 4/18/94
.\"
.Dd June 18, 2004
.Dt PAX 1
.Os
.Sh NAME
.Nm pax
.Nd read and write file archives and copy directory hierarchies
.Sh SYNOPSIS
.Nm
.Op Fl 0cdjnvzO
.Bk -words
.Op Fl E Ar limit
.Ek
.Bk -words
.Op Fl f Ar archive
.Ek
.Bk -words
.Op Fl N Ar dbdir
.Ek
.Bk -words
.Op Fl s Ar replstr
.Ar ...\&
.Ek
.Bk -words
.Op Fl U Ar user
.Ar ...\&
.Ek
.Bk -words
.Op Fl G Ar group
.Ar ...\&
.Ek
.Bk -words
.Oo
.Fl T
.Op Ar from_date
.Sm off
.Oo , Ar to_date Oc
.Sm on
.Oc
.Ar ...\&
.Ek
.Op Ar pattern ...\&
.Nm
.Fl r
.Op Fl cdijknuvzADOYZ
.Bk -words
.Op Fl E Ar limit
.Ek
.Bk -words
.Op Fl f Ar archive
.Ek
.Bk -words
.Op Fl N Ar dbdir
.Ek
.Bk -words
.Op Fl o Ar options
.Ar ...\&
.Ek
.Bk -words
.Op Fl p Ar string
.Ar ...\&
.Ek
.Bk -words
.Op Fl s Ar replstr
.Ar ...\&
.Ek
.Bk -words
.Op Fl U Ar user
.Ar ...\&
.Ek
.Bk -words
.Op Fl G Ar group
.Ar ...\&
.Ek
.Bk -words
.Oo
.Fl T
.Op Ar from_date
.Sm off
.Oo , Ar to_date Oc
.Sm on
.Oc
.Ar ...\&
.Ek
.Op Ar pattern ...\&
.Nm
.Fl w
.Op Fl dijtuvzAHLMOPX
.Bk -words
.Op Fl b Ar blocksize
.Ek
.Oo
.Op Fl a
.Op Fl f Ar archive
.Oc
.Bk -words
.Op Fl x Ar format
.Ek
.Bk -words
.Op Fl B Ar bytes
.Ek
.Bk -words
.Op Fl N Ar dbdir
.Ek
.Bk -words
.Op Fl o Ar options
.Ar ...\&
.Ek
.Bk -words
.Op Fl s Ar replstr
.Ar ...\&
.Ek
.Bk -words
.Op Fl U Ar user
.Ar ...\&
.Ek
.Bk -words
.Op Fl G Ar group
.Ar ...\&
.Ek
.Bk -words
.Oo
.Fl T
.Op Ar from_date
.Sm off
.Oo , Ar to_date Oc
.Oo /[ Cm c ] [ Cm m ] Oc
.Sm on
.Oc
.Ar ...\&
.Ek
.Op Ar file ...\&
.Nm
.Fl r
.Fl w
.Op Fl dijklntuvzADHLMOPXYZ
.Bk -words
.Op Fl N Ar dbdir
.Ek
.Bk -words
.Op Fl p Ar string
.Ar ...\&
.Ek
.Bk -words
.Op Fl s Ar replstr
.Ar ...\&
.Ek
.Bk -words
.Op Fl U Ar user
.Ar ...\&
.Ek
.Bk -words
.Op Fl G Ar group
.Ar ...\&
.Ek
.Bk -words
.Oo
.Fl T
.Op Ar from_date
.Sm off
.Oo , Ar to_date Oc
.Oo /[ Cm c ] [ Cm m ] Oc
.Sm on
.Oc
.Ar ...\&
.Ek
.Bk -words
.Op Ar file ...\&
.Ek
.Ar directory
.Sh DESCRIPTION
.Nm
will read, write, and list the members of an archive file,
and will copy directory hierarchies.
If the archive file is of the form:
.Ar [[user@]host:]file
then the archive will be processed using
.Xr rmt 8 .
.Pp
.Nm
operation is independent of the specific archive format,
and supports a wide variety of different archive formats.
A list of supported archive formats can be found under the description of the
.Fl x
option.
.Pp
The presence of the
.Fl r
and the
.Fl w
options specifies which of the following functional modes
.Nm
will operate under:
.Em list , read , write ,
and
.Em copy .
.Bl -tag -width 6n
.It Aq none
.Em List .
.Nm
will write to
.Dv standard output
a table of contents of the members of the archive file read from
.Dv standard input ,
whose pathnames match the specified
.Ar patterns .
The table of contents contains one filename per line
and is written using single line buffering.
.It Fl r
.Em Read .
.Nm
extracts the members of the archive file read from the
.Dv standard input ,
with pathnames matching the specified
.Ar patterns .
The archive format and blocking is automatically determined on input.
When an extracted file is a directory, the entire file hierarchy
rooted at that directory is extracted.
All extracted files are created relative to the current file hierarchy.
The setting of ownership, access and modification times, and file mode of
the extracted files are discussed in more detail under the
.Fl p
option.
.It Fl w
.Em Write .
.Nm
writes an archive containing the
.Ar file
operands to
.Dv standard output
using the specified archive format.
When no
.Ar file
operands are specified, a list of files to copy with one per line is read from
.Dv standard input .
When a
.Ar file
operand is also a directory, the entire file hierarchy rooted
at that directory will be included.
.It Fl r Fl w
.Em Copy .
.Nm
copies the
.Ar file
operands to the destination
.Ar directory .
When no
.Ar file
operands are specified, a list of files to copy with one per line is read from
the
.Dv standard input .
When a
.Ar file
operand is also a directory the entire file
hierarchy rooted at that directory will be included.
The effect of the
.Em copy
is as if the copied files were written to an archive file and then
subsequently extracted, except that there may be hard links between
the original and the copied files (see the
.Fl l
option below).
.Pp
.Em Warning :
The destination
.Ar directory
must not be one of the
.Ar file
operands or a member of a file hierarchy rooted at one of the
.Ar file
operands.
The result of a
.Em copy
under these conditions is unpredictable.
.El
.Pp
While processing a damaged archive during a
.Em read
or
.Em list
operation,
.Nm
will attempt to recover from media defects and will search through the archive
to locate and process the largest number of archive members possible (see the
.Fl E
option for more details on error handling).
.Sh OPERANDS
The
.Ar directory
operand specifies a destination directory pathname.
If the
.Ar directory
operand does not exist, or it is not writable by the user,
or it is not of type directory,
.Nm
will exit with a non-zero exit status.
.Pp
The
.Ar pattern
operand is used to select one or more pathnames of archive members.
Archive members are selected using the pattern matching notation described
by
.Xr fnmatch 3 .
When the
.Ar pattern
operand is not supplied, all members of the archive will be selected.
When a
.Ar pattern
matches a directory, the entire file hierarchy rooted at that directory will
be selected.
When a
.Ar pattern
operand does not select at least one archive member,
.Nm
will write these
.Ar pattern
operands in a diagnostic message to
.Dv standard error
and then exit with a non-zero exit status.
.Pp
The
.Ar file
operand specifies the pathname of a file to be copied or archived.
When a
.Ar file
operand does not select at least one archive member,
.Nm
will write these
.Ar file
operand pathnames in a diagnostic message to
.Dv standard error
and then exit with a non-zero exit status.
.Sh OPTIONS
The following options are supported:
.Bl -tag -width 4n
.It Fl r
Read an archive file from
.Dv standard input
and extract the specified
.Ar files .
If any intermediate directories are needed in order to extract an archive
member, these directories will be created as if
.Xr mkdir 2
was called with the bitwise inclusive
.Dv OR
of
.Dv S_IRWXU , S_IRWXG ,
and
.Dv S_IRWXO
as the mode argument.
When the selected archive format supports the specification of linked
files and these files cannot be linked while the archive is being extracted,
.Nm
will write a diagnostic message to
.Dv standard error
and exit with a non-zero exit status at the completion of operation.
.It Fl w
Write files to the
.Dv standard output
in the specified archive format.
When no
.Ar file
operands are specified,
.Dv standard input
is read for a list of pathnames with one per line without any leading or
trailing
.Aq blanks .
.It Fl a
Append
.Ar files
to the end of an archive that was previously written.
If an archive format is not specified with a
.Fl x
option, the format currently being used in the archive will be selected.
Any attempt to append to an archive in a format different from the
format already used in the archive will cause
.Nm
to exit immediately
with a non-zero exit status.
The blocking size used in the archive volume where writing starts
will continue to be used for the remainder of that archive volume.
.Pp
.Em Warning :
Many storage devices are not able to support the operations necessary
to perform an append operation.
Any attempt to append to an archive stored on such a device may damage the
archive or have other unpredictable results.
Tape drives in particular are more likely to not support an append operation.
An archive stored in a regular file system file or on a disk device will
usually support an append operation.
.It Fl b Ar blocksize
When
.Em writing
an archive,
block the output at a positive decimal integer number of
bytes per write to the archive file.
The
.Ar blocksize
must be a multiple of 512 bytes with a maximum of 32256 bytes.
A
.Ar blocksize
can end with
.Li k
or
.Li b
to specify multiplication by 1024 (1K) or 512, respectively.
A pair of
.Ar blocksizes
can be separated by
.Li x
to indicate a product.
A specific archive device may impose additional restrictions on the size
of blocking it will support.
When blocking is not specified, the default
.Ar blocksize
is dependent on the specific archive format being used (see the
.Fl x
option).
.It Fl c
Match all file or archive members
.Em except
those specified by the
.Ar pattern
and
.Ar file
operands.
.It Fl d
Cause files of type directory being copied or archived, or archive members of
type directory being extracted, to match only the directory file or archive
member and not the file hierarchy rooted at the directory.
.It Fl f Ar archive
Specify
.Ar archive
as the pathname of the input or output archive, overriding the default
.Dv standard input
(for
.Em list
and
.Em read )
or
.Dv standard output
(for
.Em write ) .
A single archive may span multiple files and different archive devices.
When required,
.Nm
will prompt for the pathname of the file or device of the next volume in the
archive.
.It Fl i
Interactively rename files or archive members.
For each archive member matching a
.Ar pattern
operand or each file matching a
.Ar file
operand,
.Nm
will prompt to
.Pa /dev/tty
giving the name of the file, its file mode and its modification time.
.Nm
will then read a line from
.Pa /dev/tty .
If this line is blank, the file or archive member is skipped.
If this line consists of a single period, the
file or archive member is processed with no modification to its name.
Otherwise, its name is replaced with the contents of the line.
.Nm
will immediately exit with a non-zero exit status if
.Aq Dv EOF
is encountered when reading a response or if
.Pa /dev/tty
cannot be opened for reading and writing.
.It Fl j
Use
.Xr bzip2 1
for compression when reading or writing archive files.
.It Fl k
Do not overwrite existing files.
.It Fl l
Link files.
(The letter ell).
In the
.Em copy
mode
.Fl ( r
.Fl w ) ,
hard links are made between the source and destination file hierarchies
whenever possible.
.It Fl n
Select the first archive member that matches each
.Ar pattern
operand.
No more than one archive member is matched for each
.Ar pattern .
When members of type directory are matched, the file hierarchy rooted at that
directory is also matched (unless
.Fl d
is also specified).
.It Fl o Ar options
Information to modify the algorithm for extracting or writing archive files
which is specific to the archive format specified by
.Fl x .
In general,
.Ar options
take the form:
.Cm name=value
.It Fl p Ar string
Specify one or more file characteristic options (privileges).
The
.Ar string
option-argument is a string specifying file characteristics to be retained or
discarded on extraction.
The string consists of the specification characters
.Cm a , e , f ,
.Cm m , o ,
and
.Cm p .
Multiple characteristics can be concatenated within the same string
and multiple
.Fl p
options can be specified.
The meaning of the specification characters are as follows:
.Bl -tag -width 2n
.It Cm a
Do not preserve file access times.
By default, file access times are preserved whenever possible.
.It Cm e
.Sq Preserve everything ,
the user ID, group ID, file mode bits,
file access time, and file modification time.
This is intended to be used by
.Em root ,
someone with all the appropriate privileges, in order to preserve all
aspects of the files as they are recorded in the archive.
The
.Cm e
flag is the sum of the
.Cm o
and
.Cm p
flags.
.\" .It Cm f
.\" Do not preserve file flags.
.\" By default, file flags are preserved whenever possible.
.It Cm m
Do not preserve file modification times.
By default, file modification times are preserved whenever possible.
.It Cm o
Preserve the user ID and group ID.
.It Cm p
.Sq Preserve
the file mode bits.
This is intended to be used by a
.Em user
with regular privileges who wants to preserve all aspects of the file other
than the ownership.
The file times are preserved by default, but two other flags are offered to
disable this and use the time of extraction instead.
.El
.Pp
In the preceding list,
.Sq preserve
indicates that an attribute stored in the archive is given to the
extracted file, subject to the permissions of the invoking
process.
Otherwise the attribute of the extracted file is determined as
part of the normal file creation action.
If neither the
.Cm e
nor the
.Cm o
specification character is specified, or the user ID and group ID are not
preserved for any reason,
.Nm
will not set the
.Dv S_ISUID
.Em ( setuid )
and
.Dv S_ISGID
.Em ( setgid )
bits of the file mode.
If the preservation of any of these items fails for any reason,
.Nm
will write a diagnostic message to
.Dv standard error .
Failure to preserve these items will affect the final exit status,
but will not cause the extracted file to be deleted.
If the file characteristic letters in any of the string option-arguments are
duplicated or conflict with each other, the one(s) given last will take
precedence.
For example, if
.Dl Fl p Ar eme
is specified, file modification times are still preserved.
.It Fl s Ar replstr
Modify the file or archive member names specified by the
.Ar pattern
or
.Ar file
operands according to the substitution expression
.Ar replstr ,
using the syntax of the
.Xr ed 1
utility regular expressions.
The format of these regular expressions are:
.Dl /old/new/[gp]
As in
.Xr ed 1 ,
.Cm old
is a basic regular expression and
.Cm new
can contain an ampersand (\*[Am]), \\n (where n is a digit) back-references,
or subexpression matching.
The
.Cm old
string may also contain
.Aq Dv newline
characters.
Any non-null character can be used as a delimiter (/ is shown here).
Multiple
.Fl s
expressions can be specified.
The expressions are applied in the order they are specified on the
command line, terminating with the first successful substitution.
The optional trailing
.Cm g
continues to apply the substitution expression to the pathname substring
which starts with the first character following the end of the last successful
substitution.
The first unsuccessful substitution stops the operation of the
.Cm g
option.
The optional trailing
.Cm p

archivers/pax/files/pax.c
/*	$NetBSD: pax.c,v 1.8 2005/12/01 03:00:01 minskim Exp $	*/

/*-
 * Copyright (c) 1992 Keith Muller.
 * Copyright (c) 1992, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Keith Muller of the University of California, San Diego.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#if HAVE_CONFIG_H
#include "config.h"
#endif
#if HAVE_NBTOOL_CONFIG_H
#include "nbtool_config.h"
#endif

#include <nbcompat.h>
#if HAVE_SYS_CDEFS_H
#include <sys/cdefs.h>
#endif
#if !defined(lint)
__COPYRIGHT("@(#) Copyright (c) 1992, 1993\n\
	The Regents of the University of California.  All rights reserved.\n");
#if 0
static char sccsid[] = "@(#)pax.c	8.2 (Berkeley) 4/18/94";
#else
__RCSID("$NetBSD: pax.c,v 1.8 2005/12/01 03:00:01 minskim Exp $");
#endif
#endif /* not lint */

#if HAVE_SYS_TYPES_H
#include <sys/types.h>
#endif
#if HAVE_SYS_PARAM_H
#include <sys/param.h>
#endif
#if HAVE_SYS_STAT_H
#include <sys/stat.h>
#endif
#if HAVE_SYS_TIME_H
#include <sys/time.h>
#endif
#if HAVE_SYS_RESOURCE_H
#include <sys/resource.h>
#endif
#if HAVE_STDIO_H
#include <stdio.h>
#endif
#if HAVE_FCNTL_H
#include <fcntl.h>
#endif
#if HAVE_SIGNAL_H
#include <signal.h>
#endif
#if HAVE_UNISTD_H
#include <unistd.h>
#endif
#if HAVE_STDLIB_H
#include <stdlib.h>
#endif
#if HAVE_STRING_H
#include <string.h>
#endif
#if HAVE_ERRNO_H
#include <errno.h>
#endif
#if HAVE_PATHS_H
#include <paths.h>
#endif
#include "pax.h"
#include "extern.h"
static int gen_init(void);

/*
 * PAX main routines, general globals and some simple start up routines
 */

/*
 * Variables that can be accessed by any routine within pax
 */
int	act = ERROR;		/* read/write/append/copy */
FSUB	*frmt = NULL;		/* archive format type */
int	cflag;			/* match all EXCEPT pattern/file */
int	cwdfd = -1;		/* starting cwd */
int	dflag;			/* directory member match only  */
int	iflag;			/* interactive file/archive rename */
int	jflag;			/* pass through bzip2 */
int	kflag;			/* do not overwrite existing files */
int	lflag;			/* use hard links when possible */
int	nflag;			/* select first archive member match */
int	tflag;			/* restore access time after read */
int	uflag;			/* ignore older modification time files */
int	vflag;			/* produce verbose output */
int	zflag;			/* use gzip */
int	Aflag;			/* honor absolute path */
int	Dflag;			/* same as uflag except inode change time */
int	Hflag;			/* follow command line symlinks (write only) */
int	Lflag;			/* follow symlinks when writing */
int	Mflag;			/* treat stdin as an mtree(8) specfile */
int	Xflag;			/* archive files with same device id only */
int	Yflag;			/* same as Dflg except after name mode */
int	Zflag;			/* same as uflg except after name mode */
int	vfpart;			/* is partial verbose output in progress */
int	patime = 1;		/* preserve file access time */
int	pmtime = 1;		/* preserve file modification times */
int	nodirs;			/* do not create directories as needed */
int	pfflags = 1;		/* preserve file flags */
int	pmode;			/* preserve file mode bits */
int	pids;			/* preserve file uid/gid */
int	rmleadslash = 0;	/* remove leading '/' from pathnames */
int	exit_val;		/* exit value */
int	docrc;			/* check/create file crc */
int	to_stdout;		/* extract to stdout */
char	*dirptr;		/* destination dir in a copy */
char	*ltmfrmt;		/* -v locale time format (if any) */
const char *argv0;		/* root of argv[0] */
sigset_t s_mask;		/* signal mask for cleanup critical sect */
FILE	*listf;			/* file pointer to print file list to */
char	*tempfile;		/* tempfile to use for mkstemp(3) */
char	*tempbase;		/* basename of tempfile to use for mkstemp(3) */
int	forcelocal;		/* force local operation even if the name 
				 * contains a :
				 */
int	secure = 1;		/* don't extract names that contain .. */

/*
 *	PAX - Portable Archive Interchange
 *
 *	A utility to read, write, and write lists of the members of archive
 *	files and copy directory hierarchies. A variety of archive formats
 *	are supported (some are described in POSIX 1003.1 10.1):
 *
 *		ustar - 10.1.1 extended tar interchange format
 *		cpio  - 10.1.2 extended cpio interchange format
 *		tar - old BSD 4.3 tar format
 *		binary cpio - old cpio with binary header format
 *		sysVR4 cpio -  with and without CRC
 *
 * This version is a superset of IEEE Std 1003.2b-d3
 *
 * Summary of Extensions to the IEEE Standard:
 *
 * 1	READ ENHANCEMENTS
 * 1.1	Operations which read archives will continue to operate even when
 *	processing archives which may be damaged, truncated, or fail to meet
 *	format specs in several different ways. Damaged sections of archives
 *	are detected and avoided if possible. Attempts will be made to resync
 *	archive read operations even with badly damaged media.
 * 1.2	Blocksize requirements are not strictly enforced on archive read.
 *	Tapes which have variable sized records can be read without errors.
 * 1.3	The user can specify via the non-standard option flag -E if error
 *	resync operation should stop on a media error, try a specified number
 *	of times to correct, or try to correct forever.
 * 1.4	Sparse files (lseek holes) stored on the archive (but stored with blocks
 *	of all zeros will be restored with holes appropriate for the target
 *	filesystem
 * 1.5	The user is notified whenever something is found during archive
 *	read operations which violates spec (but the read will continue).
 * 1.6	Multiple archive volumes can be read and may span over different
 *	archive devices
 * 1.7	Rigidly restores all file attributes exactly as they are stored on the
 *	archive.
 * 1.8	Modification change time ranges can be specified via multiple -T
 *	options. These allow a user to select files whose modification time
 *	lies within a specific time range.
 * 1.9	Files can be selected based on owner (user name or uid) via one or more
 *	-U options.
 * 1.10	Files can be selected based on group (group name or gid) via one o
 *	more -G options.
 * 1.11	File modification time can be checked against existing file after
 *	name modification (-Z)
 *
 * 2	WRITE ENHANCEMENTS
 * 2.1	Write operation will stop instead of allowing a user to create a flawed
 *	flawed archive (due to any problem).
 * 2.2	Archives written by pax are forced to strictly conform to both the
 *	archive and pax the spceific format specifications.
 * 2.3	Blocking size and format is rigidly enforced on writes.
 * 2.4	Formats which may exhibit header overflow problems (they have fields
 *	too small for large file systems, such as inode number storage), use
 *	routines designed to repair this problem. These techniques still
 *	conform to both pax and format specifications, but no longer truncate
 *	these fields. This removes any restrictions on using these archive
 *	formats on large file systems.
 * 2.5	Multiple archive volumes can be written and may span over different
 *	archive devices
 * 2.6	A archive volume record limit allows the user to specify the number
 *	of bytes stored on an archive volume. When reached the user is
 *	prompted for the next archive volume. This is specified with the
 *	non-standard -B flag. The limit is rounded up to the next blocksize.
 * 2.7	All archive padding during write use zero filled sections. This makes
 *	it much easier to pull data out of flawed archive during read
 *	operations.
 * 2.8	Access time reset with the -t applies to all file nodes (including
 *	directories).
 * 2.9	Symbolic links can be followed with -L (optional in the spec).
 * 2.10	Modification or inode change time ranges can be specified via
 *	multiple -T options. These allow a user to select files whose
 *	modification or inode change time lies within a specific time range.
 * 2.11	Files can be selected based on owner (user name or uid) via one or more
 *	-U options.
 * 2.12	Files can be selected based on group (group name or gid) via one o
 *	more -G options.
 * 2.13	Symlinks which appear on the command line can be followed (without
 *	following other symlinks; -H flag)
 *
 * 3	COPY ENHANCEMENTS
 * 3.1	Sparse files (lseek holes) can be copied without expanding the holes
 *	into zero filled blocks. The file copy is created with holes which are
 *	appropriate for the target filesystem
 * 3.2	Access time as well as modification time on copied file trees can be
 *	preserved with the appropriate -p options.
 * 3.3	Access time reset with the -t applies to all file nodes (including
 *	directories).
 * 3.4	Symbolic links can be followed with -L (optional in the spec).
 * 3.5	Modification or inode change time ranges can be specified via
 *	multiple -T options. These allow a user to select files whose
 *	modification or inode change time lies within a specific time range.
 * 3.6	Files can be selected based on owner (user name or uid) via one or more
 *	-U options.
 * 3.7	Files can be selected based on group (group name or gid) via one o
 *	more -G options.
 * 3.8	Symlinks which appear on the command line can be followed (without
 *	following other symlinks; -H flag)
 * 3.9	File inode change time can be checked against existing file before
 *	name modification (-D)
 * 3.10	File inode change time can be checked against existing file after
 *	name modification (-Y)
 * 3.11	File modification time can be checked against existing file after
 *	name modification (-Z)
 *
 * 4	GENERAL ENHANCEMENTS
 * 4.1	Internal structure is designed to isolate format dependent and
 *	independent functions. Formats are selected via a format driver table.
 *	This encourages the addition of new archive formats by only having to
 *	write those routines which id, read and write the archive header.
 */

/*
 * main()
 *	parse options, set up and operate as specified by the user.
 *	any operational flaw will set exit_val to non-zero
 * Return: 0 if ok, 1 otherwise
 */

int
main(int argc, char **argv)
{
	const char *tmpdir;
	size_t tdlen;

	setprogname(argv[0]);

	listf = stderr;

	/*
	 * parse options, determine operational mode
	 */
	options(argc, argv);

	/*
	 * general init
	 */
	if ((gen_init() < 0) || (tty_init() < 0))
		return(exit_val);

	/*
	 * Keep a reference to cwd, so we can always come back home.
	 */
	cwdfd = open(".", O_RDONLY);
	if (cwdfd < 0) {
		syswarn(0, errno, "Can't open current working directory.");
		return(exit_val);
	}
	if (updatepath() == -1)
		return(exit_val);

	/*
	 * Where should we put temporary files?
	 */
	if ((tmpdir = getenv("TMPDIR")) == NULL || *tmpdir == '\0')
		tmpdir = _PATH_TMP;
	tdlen = strlen(tmpdir);
	while(tdlen > 0 && tmpdir[tdlen - 1] == '/')
		tdlen--;
	tempfile = malloc(tdlen + 1 + sizeof(_TFILE_BASE));
	if (tempfile == NULL) {
		tty_warn(1, "Cannot allocate memory for temp file name.");
		return(exit_val);
	}
	if (tdlen)
		memcpy(tempfile, tmpdir, tdlen);
	tempbase = tempfile + tdlen;
	*tempbase++ = '/';

	(void)time(&starttime);
#ifdef SIGINFO
	(void)signal(SIGINFO, ar_summary);
#endif
	/*
	 * select a primary operation mode
	 */
	switch(act) {
	case EXTRACT:
		extract();
		break;
	case ARCHIVE:
		archive();
		break;
	case APPND:
		if (gzip_program != NULL)
			err(1, "cannot gzip while appending");
		append();
		/* 
		 * Check if we tried to append on an empty file and
		 * turned into ARCHIVE mode.
		 */
		if (act == -ARCHIVE) {
			act = ARCHIVE;
			archive();
		}
		break;
	case COPY:
		copy();
		break;
	default:
	case LIST:
		list();
		break;
	}
	return(exit_val);
}

/*
 * sig_cleanup()
 *	when interrupted we try to do whatever delayed processing we can.
 *	This is not critical, but we really ought to limit our damage when we
 *	are aborted by the user.
 * Return:
 *	never....
 */

void
sig_cleanup(int which_sig)
{
	/*
	 * restore modes and times for any dirs we may have created
	 * or any dirs we may have read. Set vflag and vfpart so the user
	 * will clearly see the message on a line by itself.
	 */
	vflag = vfpart = 1;
#ifdef SIGXCPU
	if (which_sig == SIGXCPU)
		tty_warn(0, "CPU time limit reached, cleaning up.");
	else
#endif
		tty_warn(0, "Signal caught, cleaning up.");

	/* delete any open temporary file */
	if (xtmp_name)
		(void)unlink(xtmp_name);
	ar_close();
	proc_dir();
	if (tflag)
		atdir_end();
	exit(1);
}

/*
 * gen_init()
 *	general setup routines. Not all are required, but they really help
 *	when dealing with a medium to large sized archives.
 */

static int
gen_init(void)
{
	struct rlimit reslimit;
	struct sigaction n_hand;
	struct sigaction o_hand;

	/*
	 * Really needed to handle large archives. We can run out of memory for
	 * internal tables really fast when we have a whole lot of files...
	 */
	if (getrlimit(RLIMIT_DATA , &reslimit) == 0){
		reslimit.rlim_cur = reslimit.rlim_max;
		(void)setrlimit(RLIMIT_DATA , &reslimit);
	}

	/*
	 * should file size limits be waived? if the os limits us, this is
	 * needed if we want to write a large archive
	 */
	if (getrlimit(RLIMIT_FSIZE , &reslimit) == 0){
		reslimit.rlim_cur = reslimit.rlim_max;
		(void)setrlimit(RLIMIT_FSIZE , &reslimit);
	}

	/*
	 * increase the size the stack can grow to
	 */
	if (getrlimit(RLIMIT_STACK , &reslimit) == 0){
		reslimit.rlim_cur = reslimit.rlim_max;
		(void)setrlimit(RLIMIT_STACK , &reslimit);
	}

#ifdef RLIMIT_RSS
	/*
	 * not really needed, but doesn't hurt
	 */
	if (getrlimit(RLIMIT_RSS , &reslimit) == 0){
		reslimit.rlim_cur = reslimit.rlim_max;
		(void)setrlimit(RLIMIT_RSS , &reslimit);
	}
#endif

	/*
	 * Handle posix locale
	 *
	 * set user defines time printing format for -v option
	 */
	ltmfrmt = getenv("LC_TIME");

	/*
	 * signal handling to reset stored directory times and modes. Since
	 * we deal with broken pipes via failed writes we ignore it. We also
	 * deal with any file size limit thorugh failed writes. CPU time
	 * limits are caught and a cleanup is forced.
	 */
	if ((sigemptyset(&s_mask) < 0) || (sigaddset(&s_mask, SIGTERM) < 0) ||
	    (sigaddset(&s_mask,SIGINT) < 0)||(sigaddset(&s_mask,SIGHUP) < 0) ||
	    (sigaddset(&s_mask,SIGPIPE) < 0)||(sigaddset(&s_mask,SIGQUIT)<0)){
		tty_warn(1, "Unable to set up signal mask");
		return(-1);
	}
#ifdef SIGXCPU
	if (sigaddset(&s_mask,SIGXCPU) < 0) {
		tty_warn(1, "Unable to set up signal mask");
		return(-1);
	}
#endif
#ifdef SIGXFSZ
	if (sigaddset(&s_mask,SIGXFSZ) < 0) {
		tty_warn(1, "Unable to set up signal mask");
		return(-1);
	}
#endif

	memset(&n_hand, 0, sizeof n_hand);
	n_hand.sa_mask = s_mask;
	n_hand.sa_flags = 0;
	n_hand.sa_handler = sig_cleanup;

	if ((sigaction(SIGHUP, &n_hand, &o_hand) < 0) &&
	    (o_hand.sa_handler == SIG_IGN) &&
	    (sigaction(SIGHUP, &o_hand, &o_hand) < 0))
		goto out;

	if ((sigaction(SIGTERM, &n_hand, &o_hand) < 0) &&
	    (o_hand.sa_handler == SIG_IGN) &&
	    (sigaction(SIGTERM, &o_hand, &o_hand) < 0))
		goto out;

	if ((sigaction(SIGINT, &n_hand, &o_hand) < 0) &&
	    (o_hand.sa_handler == SIG_IGN) &&
	    (sigaction(SIGINT, &o_hand, &o_hand) < 0))
		goto out;

	if ((sigaction(SIGQUIT, &n_hand, &o_hand) < 0) &&
	    (o_hand.sa_handler == SIG_IGN) &&
	    (sigaction(SIGQUIT, &o_hand, &o_hand) < 0))
		goto out;

#ifdef SIGXCPU
	if ((sigaction(SIGXCPU, &n_hand, &o_hand) < 0) &&
	    (o_hand.sa_handler == SIG_IGN) &&
	    (sigaction(SIGXCPU, &o_hand, &o_hand) < 0))
		goto out;
#endif
	n_hand.sa_handler = SIG_IGN;
	if (sigaction(SIGPIPE, &n_hand, &o_hand) < 0)
		goto out;
#ifdef SIGXFSZ
	if (sigaction(SIGXFSZ, &n_hand, &o_hand) < 0)
		goto out;
#endif
	return(0);

    out:
	syswarn(1, errno, "Unable to set up signal handler");
	return(-1);
}
